runtime/doc/lsp.txt

```diff
@@ -1744,6 +1744,31 @@ is_enabled({filter})                         *vim.lsp.inlay_hint.is_enabled()*
         (`boolean`)
 
 
+==============================================================================
+Lua module: vim.lsp.selection_range                      *lsp-selection_range*
```

Comment:
I think this should just be a single function `vim.lsp.buf.selection_range('expand'|'shrink', opts)`.

I know it's not tied strictly to the buffer and more to a window but so are most functions in `vim.lsp.buf`.

Reply:
Alternative is to think of a more general node/semantic selection mechanism which could internally be powered by treesitter with lsp as fallback. Assuming that they behave the same with treesitter most likely having less overhead/better performance if the buffer is already parsed.

(See incremental_selection in nvim-treesitter (?))

Reply:
Why not both? We need one for the other anyway.

Reply:
I am still considering whether `vim.lsp.buf` would be more appropriate (it just needs one parameter, since `count` can represent either `expand` or `shrink`). However, if `selection_range` is treated as a separate module, adding new APIs should be convenient, such as `vim.lsp.selection_range.get({pos})`. This could provide some extensibility.

Reply:
It would be great if we could use the tree-sitter at the same time. However, I think it might be better to have a dedicated module for this, because `selection_range` is just part of their overlapping functionality. For example, `foldexpr` also has overlaps. If we could provide a foldexpr that uses LSP when LSP support is available, otherwise uses tree-sitter.

Semantic highlighting works well with tree-sitter highlighting in this regard. I haven't seen an editor with richer code highlighting than nvim, but highlighting only needs to be overridden according to priority, so there's no need for special implementation.

Reply:
> Why not both? We need one for the other anyway.

Sure we can have both, question is in which form.

Do we end up with:

- lsp.selection_range.expand
- lsp.selection_range.shrink
- treesitter.shrink
- treesitter.expand

+ others, like jump (`vim.jump` to next edge?) and whatever use-cases people come up

Or do we come up with an approach that composes a bit better and avoids the mirroring, reducing the surface area.


> For example, foldexpr also has overlaps. If we could provide a foldexpr that uses LSP when LSP support is available, otherwise uses tree-sitter.

If you draw the parallel to foldexpr I think you need to look at it a bit differently. `foldexpr` is the implementation, but the user interface for folds are the `z` methods like `zc`, etc.

So if we mirror the approach we'd instead end up with something like a `nodesexpr` or `selectionexpr` and have `vim.selection.shrink` / `vim.selection.expand`. Or something else altogether.


I haven't given this much thought yet, but it appears to me that there's quite a large design space and I think it warrants talking about the different options. Measure twice cut once or there's a higher risk we end up with an API that's going to be deprecated a version later.

Reply:
Could this be implemented as a "scope" textobject? (In treesitter parlance, you'd rely on `locals` queries, which we could easily upstream from nvim-treesitter for the parsers we bundle. I think I would like to keep a "lower-level" tree movement as a separate object, though, since I wouldn't like an LSP to override it ever.)

Reply:
At this point, it's not clear which implementation yields the best results, so we are very much at the stage of "let the user decide". And for some servers it might be found that the treesitter implementation is much better than LSP, so there might not be a "one size fits all".